File: 036_Rotate Image.cpp
/*
Rotate Image

Rotate an n x n 2D matrix by 90 degrees clockwise, in-place.
*/

#include <bits/stdc++.h>
using namespace std;

void rotate(vector<vector<int>>& matrix) {
    int n = matrix.size();
    // transpose
    for (int i = 0; i < n; ++i)
        for (int j = i+1; j < n; ++j)
            swap(matrix[i][j], matrix[j][i]);
    // reverse each row
    for (int i = 0; i < n; ++i)
        reverse(matrix[i].begin(), matrix[i].end());
}

int main() {
    vector<vector<int>> m = {{1,2,3},{4,5,6},{7,8,9}};
    rotate(m);
    cout << "Rotated:" << endl;
    for (auto &r: m) {
        for (int v: r) cout << v << ' ';
        cout << endl;
    }
    return 0;
}

// Time Complexity: O(n^2)
// Space Complexity: O(1)

File: 036_Rotate Image.py
"""
Rotate Image

Rotate an n x n 2D matrix by 90 degrees clockwise, in-place.
"""

from typing import List

def rotate(matrix: List[List[int]]) -> None:
    n = len(matrix)
    # transpose
    for i in range(n):
        for j in range(i+1, n):
            matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
    # reverse each row
    for i in range(n):
        matrix[i].reverse()


if __name__ == '__main__':
    m = [[1,2,3],[4,5,6],[7,8,9]]
    rotate(m)
    print('Rotated:')
    for r in m: print(r)

# Time Complexity: O(n^2)
# Space Complexity: O(1)
